### PR TITLE
CP-50444: Instrument http svr with dt

### PR DESCRIPTION
**Pull request by GabrielBuica**
Instruments functions such as:

`handle_one`
`callback`
`callback1`
`dispatch`
`jsoncallback`
`response`
...

This should help us determine where the time is spent between receiving the request and sending back a response.

Latest BVT passed: 202675 (Dev Run)

### CODE REVIEW

**Review by lindig — APPROVED** [review]

File: ocaml/libs/http-lib/http_svr.ml [hunk 1]
```diff
@@ -441,9 +453,28 @@ let request_of_bio_exn ~proxy_seen ~read_timeout ~total_timeout ~max_length bio
...
+          span
+      | Error _ ->
+          None
+    in
     let r, proxy =
       request_of_bio_exn ~proxy_seen ~read_timeout ~total_timeout ~max_length ic
     in
+    let parent_span = traceparent_of_request r in
+    let loop_span =
+      match loop_span with
```
[the remark this answers is not in the document]
Comment (GabrielBuica) [reply ocaml/libs/http-lib/http_svr.ml]: I used `Option.fold` as I need the new span and the changes are not on mutable fields.

File: ocaml/libs/tracing/tracing.ml [hunk 2]
```diff
@@ -169,7 +169,7 @@ module SpanContext = struct
 end
 
 module SpanLink = struct
-  type t = {context: SpanContext.t; attributes: (string * string) list}
```
Comment (psafont) [inline ocaml/libs/tracing/tracing.ml:170]: Can you remove `-warn-error -69` from the file `xen-api.git/dune`? This transforms these warnings into errors, which makes it very difficult to reintroduce these

**Comment by psafont** [comment]
> This should fix the issue until we implement a better abstraction for the span hashtables.

Also known as 'never' ;)

**Comment by GabrielBuica** [comment]
image
 For reference, this is how the trace looks.

File: ocaml/libs/http-lib/dune [hunk 3]
```diff
@@ -45,6 +45,7 @@
     ipaddr
     polly
     threads.posix
+    tracing
```
Comment (psafont) [inline ocaml/libs/http-lib/dune:48]: All packages using this library need to have their opam dependencies checked: they will need to add `xapi-tracing` to them:

```
ocaml/database/dune
36:    httpsvr
80:    httpsvr
```
This is used only by binaries in the package xapi

```
ocaml/libs/http-lib/dune
38:  (name httpsvr)
127:    httpsvr
```
same, for testing the xapi package.

```
ocaml/xcp-rrdd/bin/rrdd/dune
12:    httpsvr
51:    httpsvr
```
This are in the xapi-rrdd package.

```
ocaml/tests/dune
21:    httpsvr
66:    httpsvr
```
These are tests in the xapi package.

```
ocaml/xapi-cli-server/dune
8:    httpsvr
```
This library is only used in the xapi package.

```
ocaml/xapi/dune
64:    httpsvr
118:    httpsvr
231:    httpsvr
```
These are all part of the xapi package.

So two packages are affected by this change, xapi and xapi-rrdd. The first already declares tracing as a dependency, but xapi-rrdd doesn't. It's definition needs to be change to depend on xapi-tracing, this is in dune-project. Remember to run make as well: the corresponding xapi-rrdd will be modified accordingly by dune.

**Review by psafont — CHANGES_REQUESTED** [review]

File: ocaml/libs/http-lib/dune (hunk 3 above)
Comment (GabrielBuica) [reply ocaml/libs/http-lib/dune:48]: This https://github.com/xapi-project/xen-api/pull/5888/commits/9e5acaa43f55e99e25210e7002b0a462024ea244?

Comment (psafont) [reply ocaml/libs/http-lib/dune:48]: yeah, that works

**Review by psafont — APPROVED** [review]

**Comment by GabrielBuica** [comment]
Passes BVT 202354.

**Comment by GabrielBuica** [comment]
Last commit updates the parent so that the trace looks like this:

image

File: ocaml/libs/tracing/tracing.ml [hunk 4]
```diff
@@ -624,6 +626,24 @@ module Tracer = struct
       let span = Span.start ~attributes ~name ~parent ~span_kind () in
       Spans.add_to_spans ~span ; Ok (Some span)
 
+  let update_span_with_parent span (parent : Span.t option) =
```
Comment (mg12) [inline ocaml/libs/tracing/tracing.ml:629]: as this is changing the underlying data structure, we probably want to guard it with the observe flag as in https://github.com/xapi-project/xen-api/blob/master/ocaml/libs/tracing/tracing.ml#L656 so thta it's a noop when tracing is not enabled

File: ocaml/libs/tracing/tracing.ml [hunk 5]
```diff
@@ -624,6 +626,24 @@ module Tracer = struct
...
+        Some span
+    | Some parent ->
+        let _ : Span.t option = Spans.remove_from_spans span in
+        let old_context = Span.get_context span in
+        let new_context : SpanContext.t =
+          SpanContext.context
+            (SpanContext.trace_id_of_span_context parent.context)
+            old_context.span_id
+        in
+        let updated_span = {span with parent= Some parent} in
```
Comment (mg12) [inline ocaml/libs/tracing/tracing.ml:641]: it may make sense to use existing_span here instead of span?

File: ocaml/libs/tracing/tracing.ml (hunk 4 above)
Comment (GabrielBuica) [reply ocaml/libs/tracing/tracing.ml:629]: https://github.com/xapi-project/xen-api/pull/5888/commits/28207c0632ba6492acf1bef49e30e964c864b76a

File: ocaml/libs/tracing/tracing.ml [hunk 6]
```diff
@@ -624,6 +626,24 @@ module Tracer = struct
       let span = Span.start ~attributes ~name ~parent ~span_kind () in
       Spans.add_to_spans ~span ; Ok (Some span)
 
+  let update_span_with_parent span (parent : Span.t option) =
+    match parent with
+    | None ->
+        Some span
+    | Some parent ->
+        let _ : Span.t option = Spans.remove_from_spans span in
```
[the remark this answers is not in the document]
Comment (GabrielBuica) [reply ocaml/libs/tracing/tracing.ml]: https://github.com/xapi-project/xen-api/pull/5888/commits/28207c0632ba6492acf1bef49e30e964c864b76a

File: ocaml/libs/tracing/tracing.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (GabrielBuica) [reply ocaml/libs/tracing/tracing.ml]: https://github.com/xapi-project/xen-api/pull/5888/commits/28207c0632ba6492acf1bef49e30e964c864b76a

**Review by mg12 — APPROVED** [review]

**Pull request merged**